File: model/bill.rb
require_relative 'application_model'
require_relative 'vote'
require_relative 'legislator'

class Bill < ApplicationModel
  attr_reader :id, :title, :sponsor_id

  def self.filename
    'bills'
  end

  def self.all
    read.map { |row| new(row[:id], row[:title], row[:sponsor_id]) }
  end

  def initialize(id, title, sponsor_id)
    super()
    @id = id
    @title = title
    @sponsor_id = sponsor_id
  end

  def vote
    @vote ||= Vote.find_by(column: :bill_id, value: id)
  end

  def supports
    vote.supports
  end

  def oppositions
    vote.oppositions
  end

  def primary_sponsor
    Legislator.find(sponsor_id)&.name || 'Unknown'
  end
end

File: model/bill.js
import ApplicationModel from "./application_model.js";
import Legislator from "./legislator.js";
import Vote from "./vote.js";

export default class Bill extends ApplicationModel {
  static filename = "bills";

  static async all() {
    return this.read().then((array) =>
      array.map((row) => new this(row.id, row.title, row.sponsor_id))
    );
  }

  constructor(id, title, sponsor_id) {
    super();
    this.id = id;
    this.title = title;
    this.sponsor_id = sponsor_id;
  }

  async vote() {
    const votes = await Vote.all();
    return votes.find((vote) => vote.bill_id === this.id);
  }

  async supports() {
    return this.vote().then((vote) => vote.supports());
  }

  async oppositions() {
    return this.vote().then((vote) => vote.oppositions());
  }

  async primary_sponsor() {
    const legislator = await Legislator.find(this.sponsor_id);
    return legislator?.name || "Unknown";
  }
}
